Theory_assignment_two/assignment_2.md

```diff
@@ -0,0 +1,453 @@
+
+High-level Design(HLD)
+
+- Description and name of each module
+- Outline of functionality of every module
+- Interface relationship and dependencies between modules
+- Database tables identified along with key elements
+- Complete architecture diagrams with technology details
+-
+-
+Low-Level Design (LLD)
+
+- Functional logic of modules
+- Complete detail of interface
+- Dealing with all types of dependency issues
+- Inputs and outputs for every module
+- Listing of error messages
+
+
+4) Implementation
+
+This is where the developers code using chosen programming languages. Tasks are devided into units or modules and assigned to different developers. The implementation phase os the longest phase of the SDLC process.
+
+
+5) Testing
+
+Once the software is completed it is deployed in the testing environment. The testing team check the functionality of the entire system. This is to verify that the entire application works to the customers needs. Testing involves:
+
+-Unit Testing
+-Module Testing
+-Sub-system Testing
+-System Testing
+-Acceptance Testing
+
+
+6) Evaluation
+
+Once a system has been delivered to the end user/client it undergoes a post-implementation review. The delivered system should then fully evaluate on the basis of its
+-effectiveness, usability and maintainability
+
+
+7) Maintenance 
+
+Once the software is deployed and is used by customers.
+
+The software will be maintained by fixing bugs, undergoing upgrades and enhancement through the addition of new features.
+
+
+## 6.What's the difference between agile and waterfall? 
+
+|     Agile                                                      |     Waterfall                                    |   |   |   |
+|----------------------------------------------------------------|--------------------------------------------------|---|---|---|
+|     Multiple lifecycle phases can run in   parallel            |     Isolated phases that flow into each other    |   |   |   |
+|     Continuous iteration of development and   testing          |     Structured software development process      |   |   |   |
+|     Not as useful for small development   projects             |     Better suited for smaller sized projects     |   |   |   |
+|     Incremental approach                                       |     Sequential design process                    |   |   |   |
+|     Testing occurs concurrently with software   development    |     Testing comes after the building phase       |   |   |   |
+|                                                                |                                                  |   |   |   |
+
+
+
+
+
+
+
+## 7. What is a reduced function used for? 
+
+Reduce() is used for applying a function to an iterable and reducing it to a single cumulative value. It performs function on first two elements and repeats process until 1 value remains.
```

Comment:
6/10 this can be elaborated in more details. When explaning a function/method, would be great if you could break down the function signature i.e. the parameters and return value/type. Worth to mention other use cases to use reduce such as finding max, min, sum, etc.